apps/server/src/shared/domain/entity/task.entity.ts

```diff
@@ -99,102 +104,153 @@ export class Task extends BaseEntityWithTimestamps implements ILearnroomElement,
 		this.publicSubmissions = props.publicSubmissions || false;
 	}
 
-	isFinishedForUser(user: User): boolean {
-		return !!(this.finished?.contains(user) || this.course?.isFinished());
+	private getSubmissionItems(): Submission[] {
+		if (!this.submissions || !this.submissions.isInitialized(true)) {
+			throw new InternalServerErrorException('Submissions items are not loaded.');
+		}
+		const submissions = this.submissions.getItems();
+
+		return submissions;
+	}
+
+	private getFinishedUserIds(): EntityId[] {
+		const finishedObjectIds = this.finished.getIdentifiers('_id');
+		const finishedIds = finishedObjectIds.map((id): string => id.toString());
+
+		return finishedIds;
 	}
 
-	isDraft(): boolean {
+	private getParent(): ITaskParent | User {
+		const parent = this.lesson || this.course || this.creator;
+
+		return parent;
+	}
+
+	private getMaxSubmissions(): number {
+		const parent = this.getParent();
+		// For draft (user as parent) propaly user is not a student, but for maxSubmission one is valid result
+		const maxSubmissions = parent instanceof User ? 1 : parent.getStudentIds().length;
+
+		return maxSubmissions;
+	}
+
+	private isFinishedForUser(user: User): boolean {
+		const finishedUserIds = this.getFinishedUserIds();
+		const isUserInFinishedUser = finishedUserIds.some((finishedUserId) => finishedUserId === user.id);
+
+		const isCourseFinished = this.course ? this.course.isFinished() : false;
+
+		const isFinishedForUser = isUserInFinishedUser || isCourseFinished;
+
+		return isFinishedForUser;
+	}
+
+	public isDraft(): boolean {
 		// private can be undefined in the database
```

Comment:
Why is undefined casted to false here? If the task is not populated it's becoming a public task just by the fact that the task is not populated.

Reply:
you're right, private is not optional, and set to true by default...

Reply:
I do not implement this. It is old code. 
But in the DB the draft key is not enforced it can be null, or the key can not be exists. The typing in the entity is wrong, in context of the real DB situation. It is more a what we want in future.